### PR TITLE
feat: POST /trades

### PR DESCRIPTION
**Pull request by meelrossi**
Add new endpoint POST /trades to create a new trade with its validations

Closes #118

### CODE REVIEW

File: src/migrations/dapps/1718991797670_trades-and-asset-trades-table.ts [hunk 1]
```diff
@@ -28,9 +31,9 @@ export async function up(pgm: MigrationBuilder): Promise<void> {
       checks: { type: 'jsonb', notNull: true },
       signer: { type: 'varchar(42)', notNull: true },
       type: { type: TRADE_TYPE, notNull: true },
-      expires_at: { type: 'timestamp', notNull: true },
-      effective_since: { type: 'timestamp', notNull: true },
-      created_at: { type: 'timestamp', notNull: true, default: pgm.func('now()') }
+      expires_at: { type: 'timestamptz(3)', notNull: true },
```
Comment (LautaroPetaccio) [inline src/migrations/dapps/1718991797670_trades-and-asset-trades-table.ts:34]: Now that we know the problem was the timezone, do we need to specify the precision as well?

File: src/migrations/dapps/1718991797670_trades-and-asset-trades-table.ts [hunk 2]
```diff
@@ -58,15 +61,59 @@ export async function up(pgm: MigrationBuilder): Promise<void> {
...
+    { schema: SCHEMA, name: TRADE_ASSETS_ERC20_TABLE },
+    {
+      asset_id: {
+        type: 'uuid',
+        unique: true,
+        notNull: true,
+        references: `"${TRADE_ASSETS_TABLE}"(id)`,
+        onDelete: 'CASCADE'
+      },
+      amount: { type: 'numeric(78,0)', notNull: true }
```
Comment (LautaroPetaccio) [inline src/migrations/dapps/1718991797670_trades-and-asset-trades-table.ts:94]: The amount bust be `greater than 0` and lower or equal than the maximum value an `uint256` can represent. There's a check [here](https://stackoverflow.com/questions/50072618/how-to-create-a-uint256-in-postgresql) that might be useful for the case.

**Review by LautaroPetaccio — COMMENTED** [review]
Amazing work! It looks great! 🥇 
I've left some comments about some things to discuss or fix.

File: src/ports/trades/component.ts [hunk 3]
```diff
@@ -10,7 +21,54 @@ export function createTradesComponent(components: Pick<AppComponents, 'dappsData
     return { data: result.rows, count: result.rowCount }
   }
 
+  async function addTrade(trade: TradeCreation, signer: string) {
+    // validate expiration > today
+    if (trade.checks.expiration < Date.now()) {
```
[the remark this answers is not in the document]
Comment (meelrossi) [reply src/ports/trades/component.ts:26]: all the communication is in milliseconds. Im changing it to seconds when signing the trade in the ui, when validating the signature and when sending the transaction

File: src/migrations/dapps/1718991797670_trades-and-asset-trades-table.ts (hunk 1 above)
Comment (meelrossi) [reply src/migrations/dapps/1718991797670_trades-and-asset-trades-table.ts:34]: This gives us millisecond precision that will be the one used for all the communications with the api

File: src/ports/trades/component.ts [hunk 4]
```diff
@@ -10,7 +21,54 @@ export function createTradesComponent(components: Pick<AppComponents, 'dappsData
...
+            const insertedValue = await client.query<DBTradeAssetValue>(
+              getInsertTradeAssetValueByTypeQuery(asset, insertedAsset.rows[0].id)
+            )
+            return { ...insertedAsset.rows[0], ...insertedValue.rows[0] }
+          })
+        )
+        return fromDbTradeWithAssetsToTrade(insertedTrade.rows[0], assets)
+      },
+      e => {
+        throw new Error(e && typeof e === 'object' && 'message' in e ? (e.message as string) : 'Could not create trade')
```
[the remark this answers is not in the document]
Comment (meelrossi) [reply src/ports/trades/component.ts]: Do you have any suggestion of which error to handle here? i believe that we are covering most cases previous to the sql query execution

File: test/unit/trades-adapters.spec.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+      beforeEach(() => {
+        erc721Asset = {
+          ...dbTradeAssetWithValueWithBeneficiary,
+          asset_type: TradeAssetType.ERC721,
+          token_id: 'token_id_1'
+        }
+      })
+
+      it('should return trade asset with amount', () => {
```
Comment (LautaroPetaccio) [inline test/unit/trades-adapters.spec.ts:197]:
```suggestion
      it('should return trade asset with the token id and the beneficiary', () => {
```

File: test/unit/trades-adapters.spec.ts [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+      beforeEach(() => {
+        collectionItemAsset = {
+          ...dbTradeAssetWithValueWithBeneficiary,
+          asset_type: TradeAssetType.COLLECTION_ITEM,
+          item_id: 'item_id_1'
+        }
+      })
+
+      it('should return trade asset with amount', () => {
```
Comment (LautaroPetaccio) [inline test/unit/trades-adapters.spec.ts:220]:
```suggestion
      it('should return trade asset with the item id and the beneficiary', () => {
```

**Pull request merged**